File: 3771.total-score-of-dungeon-runs.java
#
# @lc app=leetcode id=3771 lang=java
#
# [3771] Total Score of Dungeon Runs
#

# @lc code=start
class Solution {
    public long totalScore(int hp, int[] damage, int[] requirement) {
        int n = damage.length;
        // prefix sum of damage
        long[] prefix = new long[n + 1];
        for (int i = 0; i < n; i++) {
            prefix[i + 1] = prefix[i] + damage[i];
        }
        long total = 0;
        // For each start index
        for (int start = 0; start < n; start++) {
            long curHP = hp;
            int score = 0;
            for (int i = start; i < n; i++) {
                curHP -= damage[i];
                if (curHP >= requirement[i]) {
                    score++;
                }
            }
            total += score;
        }
        return total;
    }
}
# @lc code=end

File: 3771.total-score-of-dungeon-runs.go
#
# @lc app=leetcode id=3771 lang=golang
#
# [3771] Total Score of Dungeon Runs
#
# @lc code=start
func totalScore(hp int, damage []int, requirement []int) int64 {
    // Use prefix sums to efficiently compute remaining health after entering rooms
    n := len(damage)
    prefix := make([]int, n+1)
    for i := 0; i < n; i++ {
        prefix[i+1] = prefix[i] + damage[i]
    }
    var total int64 = 0
    // For each starting room
    for start := 0; start < n; start++ {
        curr_hp := hp
        score := 0
        for i := start; i < n; i++ {
            curr_hp -= damage[i]
            if curr_hp >= requirement[i] {
                score++
            }
        }
        total += int64(score)
    }
    return total
}
# @lc code=end